core: add finalizer checks for ManagedChannels

carl-mastrangelo opened the pull request:
Cleaning up channels is something users should do.  To promote this
behavior, add a log message to indicate that the channel has not
been properly cleaned.

This change users PhantomReferences to avoid keeping the channel
alive and retaining too much memory.  Only the id and the target
are kept.  Additionally, the lost references are only checked at
JVM shutdown and on new channel creation.  This is done to avoid
Object finalizers.

The test added checks to see that the message is logged.  Since
java does not allow forcing of a GC cycle, this code is best
effort, giving up after about a second.  A custom log filter is
added to hook the log messages and check to see if the correct
one is present.  Handlers are not used because they are
hierarchical, and would be annoying to restore their state after
the test.

The other tests in the file contribute a lot of bad channels.  This
is reasonable, because they aren't real channels.  Hwoever, it does
mean that less than half of them are being cleaned up properly.
After trying to fix a few, it is too hard to do.  It would only
serve to massively complicate the tests.

Instead, this code just keeps track of how many it wasn't able to
clean up, and ignores them for the test.  They are still logged,
because really they should be closed.

ejona86 commented on core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:
```diff
@@ -575,8 +586,9 @@ private void maybeTerminateChannel() {
-      log.log(Level.FINE, "[{0}] Terminated", getLogId());
+      logger.log(Level.FINE, "[{0}] Terminated", getLogId());
       terminated = true;
+      phantom.terminated = true;
```
It'd probably be good to `phantom.clear()` and/or `refs.remove(phatom)` in this case.

carl-mastrangelo replied:
Done.

ejona86 commented on core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:
```diff
@@ -943,4 +955,71 @@ public String toString() {
+  @VisibleForTesting
+  static final class ManagedChannelPhantom extends PhantomReference<ManagedChannelImpl> {
+    private static final ReferenceQueue<ManagedChannelImpl> refQueue = initQueue();
+    // Retain the PhantomReferences so they don't get GC'd
```
Just to be clear, this means that we probably won't detect problems in Servlet shutdown or equivalent, where the ClassLoader itself is being GC'd, right?

carl-mastrangelo replied:
Technically yes.  This is a best effort.   Even then, the phantoms don't do any clean up themselves they just act as a final notice that something wrong has happened.  EJ#7 suggests that finalizers may be used legitimately in this manner.

carl-mastrangelo commented on core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:
```diff
@@ -943,4 +955,71 @@ public String toString() {
   }
+
+  @VisibleForTesting
+  static final class ManagedChannelPhantom extends PhantomReference<ManagedChannelImpl> {
```
WeakRefs are cleared on Minor GC compaction, even if the Channel is still around.  Phantoms are only cleared when their referent is GC'd.  Technically there is no need to clean a Phantom, since they don't keep their referent alive, and `get()` always returns null.

carl-mastrangelo commented on core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:
```diff
@@ -943,4 +955,71 @@ public String toString() {
+
+    ManagedChannelPhantom(ManagedChannelImpl chan) {
+      super(chan, refQueue);
+      allocationSite = new Throwable("Managed Channel Allocation Site");
```
Okay, but please reconsider it.  It is very useful in the library sense.  A users application may depend on another library which depends on gRPC.  If the intermediate library forgets to cleanup, the user application will see a spurious warning.  

Would a System property assuage your concerns?

ejona86 replied:
I'm concerned about things like #3272 . Having a system property to enable the Throwable would be fine. So it'd be off unless people were trying to debug. We should catch `SecurityException` when doing so.

carl-mastrangelo replied:
I went with a different approach.  This should keep the throwable alive for the minor GC collections. 

It isn't a disaster if the call site isn't captured, and this avoids adding configuration.

carl-mastrangelo commented on core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:
```diff
@@ -943,4 +955,71 @@ public String toString() {
+      }
+
+      try {
+        getRuntime().addShutdownHook(getThreadFactory(name, daemon).newThread(new QueueCleaner()));
```
Can you point me to where the logger is shutdown?  I was under the impression a loggers handler can be closed, but not the logger itself.

carl-mastrangelo replied:
Actually, after investigating further, you are right.  I have included a best effort mechanism to log to the console in the event of shutdown.

carl-mastrangelo commented:
Changes made, PTAL.

carl-mastrangelo commented:
@ejona86 ping

carl-mastrangelo commented:
> Let's also clear() if they call shutdownNow(). If they call shutdownNow() there's nothing more they can do to make sure the channel is cleaned up.

I've changed it so that if shutdownNow is called, it lowers the severity of the message.  It's still a bug (the channel was never properly terminated), but they did at least try to do the right thing.  

This also makes most of our tests quiet.

> I'm still uncomfortable with the Exception. It's something that can cause problems for everyone, while only potentially helping a minority. Let's leave it out for now.

I have modified this to check a system property.  My fear is that an application that may be far removed from the ManagedChannel allocation site will get nag warnings, but not have any way to know why.

The system property acts as a safety valve if it needs to be turned off, at which point we can revisit if it should be off by default.
